### ISSUE TITLE
Implementation of Siamese transformer

### ISSUE THREAD

**Issue by prajjwal1** [UKPLab/sentence-transformers#298]
Hi,

Thanks for the paper (SBERT) and the library. I'm trying to implement your paper. I'm closely referring to the SBERT figure from the paper. Here's how I've defined the model :
```
class PredictionHeadTransform(nn.Module):
    def __init__(self, config):
        super().__init__()
        self.pool = nn.AdaptiveAvgPool2d((8, 128))
        self.dense = nn.Linear(4096, len(config.id2label))

    def forward(self, features):
        features = self.pool(features)
        features = features.view(features.shape[0] // 4, -1)
        features = self.dense(features)
        return features


class SiameseTransformer(nn.Module):
    def __init__(self, args, config):
        super(SiameseTransformer, self).__init__()
        self.args = args
        self.model_a = AutoModel.from_pretrained(
            self.args.model_name, config=config, cache_dir=self.args.cache_dir
        )
        self.model_b = AutoModel.from_pretrained(self.args.model_name,
                           config=config, cache_dir=self.args.cache_dir)
        self.loss_fct = nn.CrossEntropyLoss()
        self.cls = PredictionHeadTransform(config)

        if self.args.freeze_a:
            logger.info("**** Freezing Model A ****")
            for param in self.model_a.encoder.parameters():
                param.requires_grad = False

        if self.args.freeze_b:
            logger.info("**** Freezing Model B ****")
            for param in self.model_b.encoder.parameters():
                param.requires_grad = False

    def forward(self, input_a, input_b):
        labels = input_a["labels"]
        input_a.pop("labels")
        input_b.pop("labels")
        output_a = self.model_a(**input_a)[0]  # [bs, seq_len, 768]
        output_b = self.model_a(**input_b)[0]
        concat_output = torch.cat(
            [output_a, output_b, (output_a - output_b), (output_a * output_b)]
        )
        logits = self.cls(concat_output)
        loss = self.loss_fct(logits, labels)
        return loss, logits
```
I tried with different configs (MLP, max_seq_len, freezing and non freezing, with pooling, without pooling, different concat methods you mentioned) but the model doesn't perform well. Here `input_a` and `input_b` represents `U` and `V` from the paper which in case of MNLI is hypothesis and premise seperated out and fed to both the model seperately. 

I scanned your library but it's bit hard to wrap mind completely as the model has been defined in a bit different manner. 

Could you please suggest what's off or what am I missing ? Or refer me to the code in this library that may solve the purpose. I do not want sentence embeddings per se, I want to train the siamese network like you've shown in paper.

Also, [here](https://github.com/UKPLab/sentence-transformers/issues/199) you mention the use of tied weights and how it can reduce the need to have second BERT, could you please suggest how this can be done ?

I've defined the model in such a way that it stays compatible with HF's `Trainer()` class. So you'd already know about the training code.

Could you please provide pointers ? Any suggestions, advice will be very helpful. 

Thanks!

**Comment by nreimers** [comment]
> Also, here you mention the use of tied weights and how it can reduce the need to have second BERT, could you please suggest how this can be done ?
This is usually done by having only one model (model_a) and passing both inputs through the same model.

I see in your code that you have two models defined, but only one model is used.

For your prediction head I don't know why you use AdaptiveAvgPool2d. A dense layer that maps the 3072 input dimensions to your labels should be sufficient.

**Comment by prajjwal1** [comment]
I used the pooling layer because you mentioned in the paper that you use pooling to get an input of fixed size (hence adaptive average pooling and to make the architecture a bit flexible). **Can you please provide where do I look if I want to train my own Siamese network like in the paper as you have mentioned** 

I give "hypothesis" and "premise" (seperated out, HF's GlueDataset concatenates it) and pass them to frozen BERT (two times, one for U and one for V). Is this step correct ? What I have noticed after spending a lot of time is that separated "hypothesis" and "premise" doesn't help at all. Please correct me if I'm wrong.
